File: ChuckNorrisObjC/ChuckNorrisObjC/ViewController.m
//
//  ViewController.m
//  ChuckNorrisObjC
//
//  Created by Maria Luiza Carvalho Sperotto on 23/01/20.
//  Copyright © 2020 Maria Luiza Carvalho Sperotto. All rights reserved.
//

#import "ViewController.h"
#import <ChuckNorrisObjC-Swift.h>


@interface ViewController ()

@property (strong, nonatomic) NSArray *myArray;



@end

@implementation ViewController

NSString *selectedCategory;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.tableView registerClass:UITableViewCell.class forCellReuseIdentifier:@"cell"];
    self.navigationItem.title = @"Categories";
    self.navigationController.navigationBar.prefersLargeTitles = YES;
    [self fetchCategoryJSON];
    
    
}

-(void)fetchCategoryJSON {
    NSLog(@"Fetching categories");
    NSString *urlString = @"https://api.chucknorris.io/jokes/categories";
    NSURL *url = [NSURL URLWithString:urlString];
    [[NSURLSession.sharedSession dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSString *fetch = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        NSLog(@"Fetch request: %@", fetch);
        NSError *err;
        
        NSArray *categories = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&err];
        self.myArray = categories;
        
        if (err) {
            NSLog(@"Failed to serialized into JSON: %@", err);
            return;
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.tableView reloadData];
        });
        
    }] resume];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _myArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    NSString *category = _myArray[indexPath.row];
    cell.textLabel.text = category;
    return cell;
}


- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NSString *category = _myArray[indexPath.row];
    selectedCategory = category;
    DetailViewController *detailVC = [[UIStoryboard storyboardWithName:@"Main" bundle:nil] instantiateViewControllerWithIdentifier:@"detailSegue"];
    detailVC.selectedCategory = selectedCategory;
    [self presentViewController:detailVC animated:YES completion:nil];
    
    
    

   
    
    

    
    
}


@end

File: Chuck Norris/Chuck Norris/ViewController.swift
//
//  ViewController.swift
//  Chuck Norris
//
//  Created by Maria Luiza Carvalho Sperotto on 13/11/19.
//  Copyright © 2019 Maria Luiza Carvalho Sperotto. All rights reserved.
//

import UIKit

class ViewController: UITableViewController {
    var jokes = [Joke]()
    var selectedCategory: String = ""
    var categorias = [String]()
    
    @IBAction func shuffle(_ sender: Any) {
        categorias.shuffle()
        self.tableView.reloadData()
        
        
    }
    
    @IBAction func reload(_ sender: Any) {
        categorias.sort()
        self.tableView.reloadData()
    }
    
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        title = "Choose a category"
        parse()
        
    }
    
    func parse() {
        let urlString = "https://api.chucknorris.io/jokes/categories"
        
        guard let url = URL(string: urlString) else { return }
        URLSession.shared.dataTask(with: url) { (data, response, error) in
            guard let data = data else { return }
            
            
                let json = try? JSONSerialization.jsonObject(with: data, options: [])
                self.categorias = json as! [String]
                DispatchQueue.main.async {
                    self.tableView.reloadData()
                }
                
            }.resume()
        
    }
    
    
    
    // Table functions
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return categorias.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Category", for: indexPath)
        let category = categorias[indexPath.row]
        cell.textLabel?.text = category
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let category = categorias[indexPath.row]
        selectedCategory = category
        performSegue(withIdentifier: "Detail", sender: nil)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "Detail" {
            let categoryDetail = segue.destination as! DetailViewController
            categoryDetail.selectedCategory = selectedCategory
        }
    }
    
}



